<a href="https://colab.research.google.com/github/davidmiheev/QuantQuestions/blob/main/Prob_Questions_%5BPart_1%5D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Probability questions from Quant Questions: Part 1
Here are collected problems where you are asked to find some probability or expectation value and solutions for these problems

## Increasing Dice Order I

> You roll a fair 6-sided die twice.
Calculate the probability that the value of the first roll is strictly less than the value of the second roll.


### Solution

Use Formula of Total Probability:
$$P(A)=\sum_k P(A|B_k)P(B_k)$$

Let $X_1$ - value of the first roll and $X_2$ - value of the second roll

Then we need to find probability that $X_2 > X_1$:
$$P(X_2>X_1) = \sum_{k=1}^6 P(X_2>X_1|X_1=k)P(X_1=k)$$

We know that $P(X_1=k)=1/6$ for any $k$
and $P(X_2>X_1|X_1=k)=(6-k)/6$

Now, we can substitute all values to the formula and get:
$$P(X_2>X_1) = 1/6\sum_{k=1}^6 (6-k)/6=5/12$$




In [ ]:
from fractions import Fraction

print('The answer is', Fraction('1/6')*sum(Fraction(f'{6-k}/6') for k in range(1, 7)))

The answer is 5/12


## Probability of Unfair Coin I

> You have a pile of 100 coins.
1 of the coins is an unfair coin and has heads on both sides. The remaining
99 coins are fair coins. You randomly select a coin from the pile and flip it
10 times. The coin lands heads all
10 times. Calculate the probability that the coin you selected is the unfair coin.

### Solution

Use Bayes Formula
$$P(A|B)=\frac{P(B|A)\cdot P(A)}{P(B)},$$

where A is the event "selected unfair coin", B is the event "The coin lands heads all 10 times"

Then $P(B|A) = 1, P(A)=1/100, P(B)=1/2^{10}\cdot 99/100+1/100$

So,
$$P(A|B)=1\cdot 1/100 \cdot \frac{1024\cdot 100}{1123}= \frac{1024}{1123}$$

In [ ]:
print('The answer is', Fraction('1024/1123'))

The answer is 1024/1123


## First Ace

> On average, how many cards in a normal deck of 52 playing cards do you need to flip over to observe your first ace?

### Solution 1: Python script

Let $X$ is the number of cards you need to flip over to observe first ace

We need to find expectation of $X$

Use definition of expected value:
$$E[X]=\sum_k kP(X=k)$$
where $P(X=k)= \frac{4}{52-(k-1)}\prod_{j=1}^{k-1}\frac{52-4-(j-1)}{52-(j-1)}$

See the script, that computes this sum below

In [ ]:
expected_value = Fraction('0')
prob = Fraction('1')
for k in range(1, 53):
  expected_value += Fraction(f'{4*k}/{53-k}')*prob
  prob *= Fraction(f'{49-k}/{53-k}')

print('The answer is', expected_value, 'or', float(expected_value))

The answer is 53/5 or 10.6


### Solution 2: Linearity of Expectation

Let $L_i, i=1,...,5$ is a length of interval $(A_{i-1}, A_i)$, where $A_i$ is a position of i-th ace in the deck ($A_0$ is the beginning of the deck and $A_5$ is the end of the deck. Because of symmetry we have $E[L_1] = E[L_2]=...= E[L_5]$

Also, we know that $$48=E[L]=E[\sum L_i]=\sum E[L_i] = 5E[L_1],$$
because of linearity of expectation.

So, $E[L_1] = 48/5$, then $E[X]=48/5+1=53/5=10.6$


In [ ]:
expected_value = Fraction('53/5')
print('The answer is', expected_value, 'or', float(expected_value))

The answer is 53/5 or 10.6


## High or Die

> Francisco rolls a fair die and records the value he rolls. Afterwards, he continues rolling the die until he obtains a value at least as large as the first roll. Let N be the number of rolls after the first he performs. Find
$E[N]$

### Solution

Let $X_1$ - value of the first roll.
We know that $$E[N]=E[E[N|X_1]]=\sum_{k=1}^6 1/6\cdot E[N|X_1=k],$$
where $E[N|X_1=k]=1/p$ is the mean for geometric distribution with parameter $p=(7-k)/6$. Since we continue rolling until first success: Francisco obtains a value at least as large as the first roll (which value is $k$).

So, now we can substitute all these in our formula for $E[N]$

In [ ]:
expected_value = sum(Fraction('1/6')*Fraction(f'6/{7-k}') for k in range(1, 7))
print('The answer is', expected_value, 'or', float(expected_value))

The answer is 49/20 or 2.45


## Local Maxima

> 14 pieces of paper labelled 1−14 are placed in a line at random. Spot i is a local maximum if the paper at the ith position is strictly larger than all of it's adjacent papers. Find the expected number of local maxima in the sequence. For example, with 6 numbers, 513246 has 3 local maxima at the first, third, and last spots.

### Solution
Let $X$ is a number of local maxima in the sequence.

Let $X_i, i = 1,...,14$ is a Bernoulli variable with parameter $p_i=P(\text{i-th spot is a local maximum})$

Let's calculate $p_i$ for each $i$:

*   $p$ for $X_1, X_{14}$ equals $1/2$ because we have two options here: a>b or a<b which have equal probabilities and one of them is a success
*   $p$ for other $X_i$ eqals $1/3$ because we have six $(=3!)$ options how to arrange labels and only two of them are success options: $a<b>c$ or $c<b>a$

Now, thanks to linearity of expectation, we have $E[X]=E[\sum X_i]=\sum E[X_i]=\sum p_i = 1/2 + 1/3 + ... + 1/3 +1/2=5$


In [ ]:
expected_value = Fraction('1/2') + 12*Fraction('1/3') + Fraction('1/2')
print('The answer is', expected_value, 'or', float(expected_value))

The answer is 5 or 5.0


## Expecting HTH

> On average, how many tosses of a fair coin does it take to see HTH?

### Solution

Let's consider Markov chain with 4 states: $S_0$ - is initial state, $S_1$ is "we have head after 1-st toss" or "H", $S_2$ is "we have tail after head" or "HT", $S_3$ - success state or "HTH" (absorbing state).
Now we have following transition probabilities $P(S_0|S_0)=1/2, P(S_1|S_0)=1/2, P(S_1|S_1)=1/2,$ $P(S_2|S_1)=1/2, P(S_0|S_2)=1/2, P(S_3|S_2)=1/2$, other transition probs are zero.

So, we get following linear system for expectated numbers of tosses to see HTH starting from $S_i$ state $E_i[N]$:
$$E_0[N]=1+1/2E_0[N]+1/2E_1[N]$$
$$E_1[N]=1+1/2E_1[N]+1/2E_2[N]$$
$$E_2[N]=1+1/2E_0[N]$$
The answer is value of $E_0[N]$ from solution of this system

In [ ]:
import numpy as np
from numpy.linalg import solve

A = np.array(
    [[1, -1, 0],
     [0, 1, -1],
     [-1, 0, 2]]
    )

b = np.array([2, 2, 2])
print("The answer is", int(solve(A, b)[0]))

The answer is 10


## Marble Mischief

> Sean has 200 red, 400 blue, and 600 green marbles in an urn. He draws out the marbles one-at-a-time without replacement. Find the probability that there is at least 1 blue and 1 green marble left in the urn right after the last red marble is selected.

### Solution

Let B and G represent the event that the absolute last marble in our
sequence is blue or green, respectively. Let R be the event in question. By Law of Total Probability,

$$P(R)=P(R|B)P(B)+P(R|G)P(G)$$

Considering the relation between numbers of marbles: 1:2:3, we have that $P(B)=1/3$, $P(G)=1/2$. Now let's calculate $P(R|B)$ and $P(R|G)$. Denote $P(R|B)$ as $x$ and $P(R|G)$ as $y$. By Law of Total Probability, $x≈x\cdot P(B)+1\cdot P(G)$ and $y\approx y\cdot P(G)+1\cdot P(B)$.

So, $x\approx P(G)/(1-P(B))$ and $y\approx P(B)/(1-P(G))$ and now we can calculate the answer

In [ ]:
print('The answer is', Fraction('1/3')/(1-Fraction('1/2'))*Fraction('1/2')+Fraction('1/2')/(1-Fraction('1/3'))*Fraction('1/3'))

The answer is 7/12


## Ten Ten

> Adam is rolling a fair 10−sided die. He gets to roll repeatedly and may decide when to stop at any time. If he obtains a value that is not 10 on each roll, he adds the upface to his total monetary sum. If he rolls a 10, he loses all of his money. If Adam plays optimally, he should stop once he has at least $k total. Find k

### Solution

Expected number of fails before success (Adam rolls a 10) is 9 (it is geometric distribution). Expected value of gain in each roll is $5. So, we need to stop when we have 9*5 = \$45. The answer is 45



## St. Petersburg Paradox

> Suppose you are offered to play a game where you flip a fair coin until you obtain a heads for the first time. If the first heads occurs on the nth flip, you are paid out $\$2^n$. What is the fair value of this game? If your answer is infinite, enter $−1$

### Solution

Let X is your gain in this game. $E[X] = \sum_x x\cdot P(X=x) = \sum 2^n\cdot(1/2)^n=\infty$.

So, the answer is $-1$